DolbyIO/Source/SdkAccessModule/Private/DeviceManagement.cpp

```diff
@@ -3,132 +3,279 @@
 #include "DeviceManagement.h"
 
 #include "Common.h"
-#include "Devices.h"
-#include "Events.h"
+#include "ErrorHandler.h"
+#include "SdkEventsObserver.h"
+
+#include "Misc/ScopeLock.h"
 
 namespace Dolby
 {
-	FDeviceManagement::FDeviceManagement(FDvcDeviceManagement& DeviceManagement, FSdkStatus& Status)
-	    : Status(Status), Events(MakeUnique<FEvents>(Status)),
-	      InputDevices(MakeUnique<FDevices>(FDevices::EDirection::Input, DeviceManagement, Status)),
-	      OutputDevices(MakeUnique<FDevices>(FDevices::EDirection::Output, DeviceManagement, Status))
+	inline auto ToFText(const std::string& String)
+	{
+		return FText::FromString(UTF8_TO_TCHAR(String.c_str()));
+	}
+
+#define DLB_UE_LOG_DEVICE(Type, Event, Name) \
+	UE_LOG(LogDolby, Log, TEXT(Type " device " Event ": %s"), *ToFText(Name).ToString())
+
+	FDeviceManagement::FDeviceManagement(FDvcDeviceManagement& DeviceManagement, ISdkEventsObserver& Observer,
+	                                     FHandlersMaker MakeHandler)
+	    : DeviceManagement(DeviceManagement), MakeHandler(MakeHandler), Observer(Observer)
 	{
-		InitializeDevices(DeviceManagement);
+		InitializeDevices();
 	}
 
 	FDeviceManagement::~FDeviceManagement() {}
 
-	void FDeviceManagement::SetInputDevice(const int Index)
+	void FDeviceManagement::SetInputDevice(const Index Index)
+	try
+	{
+		FScopeLock Lock{&AccessDevices};
+		const FDvcDevice& Device = GetDeviceAt(EDirection::input, Index);
+		DLB_UE_LOG_DEVICE("Input", "change requested", Device.name());
+		DeviceManagement.set_preferred_input_audio_device(Device).on_error(
+		    [this](auto&& Ex)
+		    {
+			    UpdateCurrentInputDeviceIndex();
+			    MakeHandler(__LINE__)(MoveTemp(Ex));
```

Comment:
To be perfectly honest this error handling stuff is confusing to analyze (ErrorHandler, MakeHandler, FHandlersMaker, the way this line looks, etc.). I trust it works and hopefully is nice to work with, but I just wanted to note that to me it's not easily readable.